File: scripts/extract-approved-cut-file.ts
import path from "path";
import { existsSync, promises as fs } from "fs";
import { execFile } from "child_process";
import { promisify } from "util";
import { cutRoot, workflowRoot } from "./workflow-paths";

const execFileAsync = promisify(execFile);

function parseArgs() {
  const args = process.argv.slice(2);
  let archive = "";
  let entry = "";
  let output = "";

  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === "--archive") archive = args[++i] ?? "";
    else if (args[i] === "--entry") entry = args[++i] ?? "";
    else if (args[i] === "--output") output = args[++i] ?? "";
    else throw new Error(`Unsupported argument: ${args[i]}`);
  }

  if (!archive || !entry || !output) {
    throw new Error("Required: --archive <path> --entry <archive entry> --output <folder>");
  }
  return { archive, entry, output };
}

function assertWithin(base: string, target: string, label: string) {
  const relative = path.relative(base, target);
  if (relative.startsWith("..") || path.isAbsolute(relative)) {
    throw new Error(`${label} is outside allowed root`);
  }
}

async function main() {
  const args = parseArgs();
  const root = workflowRoot;
  const archiveRoot = path.resolve(cutRoot, "_source-downloads");
  const archivePath = path.resolve(root, args.archive);
  const outputPath = path.resolve(root, args.output);

  assertWithin(archiveRoot, archivePath, "archive");
  assertWithin(cutRoot, outputPath, "output");
  if (!existsSync(archivePath)) throw new Error(`Archive not found: ${args.archive}`);

  await fs.mkdir(outputPath, { recursive: true });
  const { stdout, stderr } = await execFileAsync("7z", [
    "e",
    archivePath,
    args.entry,
    `-o${outputPath}`,
    "-y"
  ], { windowsHide: true });

  const files = (await fs.readdir(outputPath, { withFileTypes: true }))
    .filter((entry) => entry.isFile())
    .map((entry) => entry.name)
    .sort((a, b) => a.localeCompare(b, "en"));

  console.log(JSON.stringify({
    archive: args.archive,
    entry: args.entry,
    output: args.output,
    files,
    stdout: stdout.trim(),
    stderr: stderr.trim()
  }, null, 2));
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: scripts/extract-approved-cut-file.py
import argparse
import json
import os
from pathlib import Path
from zipfile import ZipFile


def within(base: Path, target: Path) -> bool:
    try:
        target.relative_to(base)
        return True
    except ValueError:
        return False


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--archive", required=True)
    parser.add_argument("--entry")
    parser.add_argument("--entry-index", type=int)
    parser.add_argument("--output", required=True)
    args = parser.parse_args()

    root = Path.cwd().resolve()
    archive_roots = [
        (root / "db image" / "cut" / "_source-downloads").resolve(),
        (root / "db image" / "output").resolve(),
    ]
    cut_root = (root / "db image" / "cut").resolve()
    archive_path = (root / args.archive).resolve()
    output_path = (root / args.output).resolve()

    if not any(within(archive_root, archive_path) for archive_root in archive_roots):
        raise SystemExit("archive is outside allowed temporary source roots")
    if not within(cut_root, output_path):
        raise SystemExit("output is outside cut root")
    if not archive_path.exists():
        raise SystemExit(f"archive not found: {args.archive}")

    if (args.entry is None) == (args.entry_index is None):
        raise SystemExit("provide exactly one of --entry or --entry-index")

    output_path.mkdir(parents=True, exist_ok=True)
    with ZipFile(archive_path) as archive:
        names = [name for name in archive.namelist() if not name.endswith("/")]
        if args.entry_index is not None:
            if args.entry_index < 0 or args.entry_index >= len(names):
                raise SystemExit(f"entry index out of range: {args.entry_index}")
            entry = names[args.entry_index]
        else:
            entry = args.entry
            if entry not in names:
                raise SystemExit(f"entry not found: {entry}")
        file_name = Path(entry).name
        target = output_path / file_name
        target.write_bytes(archive.read(entry))

    result = {
        "archive": args.archive,
        "entry": entry,
        "output": args.output,
        "file": str(target),
        "bytes": target.stat().st_size,
        "files": sorted(p.name for p in output_path.iterdir() if p.is_file()),
    }
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
